### PR TITLE
[reconfigurator] Introduce a combined `SledEditor` inside `BlueprintBuilder`

### PR DESCRIPTION
**Pull request by jgallagher**
This is a step on the road to #7078. The current `BlueprintBuilder` internals would not at all be amenable to squishing the disparate maps in `Blueprint` together, so this PR tries to rework those internals. `BlueprintBuilder` now does its own map squishing (inside `new_based_on()`), combining state+zones+disks+datasets into one helper (`SledEditor`). All modifications are preformed via those editors, then `BlueprintBuilder::build()` breaks the combined editors' results back out into four maps for `Blueprint`.

There should be only one functional change in this PR (marked with a `TODO-john`): previously, expunging a zone checked that that zone had not been modified in the current planning cycle. I'm not sure that's really necessary; it felt like some defensiveness born out of the complexity of the builder itself, maybe? I think we could put it back (inside `ZonesEditor`, presumably) if folks think it would be good to keep.

My intention was to change the tests as little as possible to ensure I didn't break anything as I was moving functionality around, so `new_based_on()` and `build()` have some arguably-pretty-gross concessions to behave exactly the way the builder did before these changes. I'd like to _remove_ those concessions, but those will be nontrivial behavioral changes that I don't want to try to roll in with all of this cleanup. I think I landed this pretty well; there are only a few expectorate changes (due to the slightly reworked `ZonesEditor` producing a different ordering for a few tests), and none in `omdb` (which is where I've often seen incidental planner changes bubble out).

Apologies for the size of the PR. I'd lightly recommend that the new `blueprint_editor/` module and its subcomponents should be pretty quickly reviewable and should be looked at first; they're _supposed_ to be simple and obvious, and are largely ported over from the prior storage / disks / datasets / zones editors. I did rework how we're handling #6645 backwards compat to try to reduce how much we need to pass `SledResources` around, so that complicates things in `DatasetsEditor` some, but not too bad IMO. (I also added a test for this, since I changed it and I don't think we had one before.) The `BlueprintBuilter` changes should then be more or less the natural way one would use a collection of `SledEditor`s without changing its API or behavior (yet!).

### CODE REVIEW

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs [hunk 1]
```diff
@@ -715,12 +888,13 @@ impl<'a> BlueprintBuilder<'a> {
                 continue;
             };
 
-            let is_expunged =
-                is_already_expunged(zone_config, state).map_err(|error| {
-                    Error::Planner(anyhow!(error).context(format!(
-                        "for sled {sled_id}, error computing zones to expunge"
-                    )))
-                })?;
+            // TODO-john we lost the check for "are we expunging a zone we
```
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:891]: I"m not sure I really understand why that was there. If we mark a zone expunged in this iteration, why wouldn't we set the disposition to expunged in the editor?

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs [hunk 2]
```diff
@@ -3003,4 +2961,92 @@ pub mod test {
 
         logctx.cleanup_successful();
     }
+
+    // This test can go away with
+    // https://github.com/oxidecomputer/omicron/issues/6645; for now, it
+    // confirms we maintain the compatibility layer it needs.
+    #[test]
+    fn test_backcompat_reuse_existing_database_dataset_ids() {
```
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:2969]: Great test!

**Review by andrewjstone — APPROVED** [review]
This is quite an improvement to the current state of affairs. I believe it will help me cleanup my branch for making disk expungement and decommissioning explicit in blueprints. 

You can take or leave my suggestions. I just want to merge this in and use it immediately.

File: nexus/reconfigurator/planning/src/blueprint_editor/sled_editor/datasets.rs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+/// situation where a dataset might have already existed in CRDB with an ID,
+/// but the blueprint system doesn't know about it. We accept a map of all
+/// existing dataset IDs, and then when determining the ID of a dataset,
+/// we'll try these in order:
+///
+/// 1. Is the dataset in our blueprint already? If so, use its ID.
+/// 2. Is the dataset in `preexisting_database_ids`? If so, use that ID.
+/// 3. Generate a new random ID.
+#[derive(Debug)]
+pub(crate) struct PreexistingDatasetIds(
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_editor/sled_editor/datasets.rs]: Renamed in 73045a7fe

File: nexus/reconfigurator/planning/src/blueprint_editor/sled_editor/datasets.rs [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        &self,
+        zpool_id: &ZpoolUuid,
+        kind: &DatasetKind,
+    ) -> Option<DatasetUuid> {
+        self.0.get(zpool_id).and_then(|by_kind| by_kind.get(kind).copied())
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct PartialDatasetConfig {
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_editor/sled_editor/datasets.rs:122]: Added along with the rework in bd661651b

File: nexus/reconfigurator/planning/src/blueprint_editor/sled_editor/datasets.rs [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+            reservation: None,
+            compression: CompressionAlgorithm::Off,
+        }
+    }
+
+    // Helper to generate a full `BlueprintDatasetConfig` from a partial config;
+    // we either look up the ID (if we're updating an existing dataset) or
+    // generate a new one via `rng`.
+    //
+    // TODO-cleanup It seems awkward we don't know whether we're updating or
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_editor/sled_editor/datasets.rs]: Reworked in bd661651b; among other things, `build()` is gone and `get_id` is now private

File: nexus/reconfigurator/planning/src/blueprint_editor/sled_editor.rs [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+        // When expunging a zone, we also expunge its datasets.
+        if let Some(dataset) = config.filesystem_dataset() {
+            // It would be strange to not find the dataset ID here; it would
+            // mean we have a zone (because we found `config`) that has a
+            // filesystem dataset, but that dataset isn't present. It _could_
+            // happen if the zone and its dataset had previously been expunged,
+            // and the dataset has since been cleaned up and removed from the
+            // blueprint entirely, but the (expunged) zone is still around.
+            if let Some(dataset_id) =
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_editor/sled_editor.rs]: This required reworking `DatasetsEditor::expunge()`, which made it more obvious that it's a little fishy. I added an example of what I think could go wrong in 8b32ca0d3, and then changed "expunge zone => expunge its datasets" to "expunge zone => if we actually expunged it, expunge its datasets" out of paranoia in 1696858a3.

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs [hunk 7]
```diff
@@ -440,30 +472,104 @@ impl<'a> BlueprintBuilder<'a> {
...
+        // `SledEditor`s.
+        let mut sled_editors = BTreeMap::new();
+        for (sled_id, zones) in &parent_blueprint.blueprint_zones {
+            // Prefer the sled state from our parent blueprint for sleds
+            // that were in it.
+            let state = match parent_blueprint.sled_state.get(sled_id).copied()
+            {
+                Some(state) => state,
+                None => {
+                    // If we have zones but no state for a sled, we assume
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:510]: I tried putting a check in `SledEditor::set_state()`; i.e., "if you're try to set the state to `Decommissioned`, the editor will fail if ... insert whatever checks we want", and it caused some test failures. Almost certainly those tests are taking invalid shortcuts; I'd like to take this as a followup.

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs [hunk 8]
```diff
@@ -546,29 +647,91 @@ impl<'a> BlueprintBuilder<'a> {
...
+                );
+            } else {
+                info!(
+                    self.log, "sled unchanged in new blueprint";
+                    "sled_id" => %sled_id,
+                    "blueprint_id" => %blueprint_id,
+                );
+            }
+        }
+        // Preserving backwards compatibility, for now: prune sled_state of any
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:701]: Yeah, definitely. I'd like to remove all of these "preserving backwards compatibility" blocks, as they all result in the four maps (potentially) having different sets of keys, which is confusing at best.

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs (hunk 1 above)
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:892]: This check was something else; it was saying "if we're about to mark a zone as expunged, first check that we didn't already edit this zone in some way in this same planning pass". I think it was to guard against the planner adding a zone and then immediately expunging it, or changing the parameters of a zone and then expunging it, or other nonsense operations.

File: nexus/reconfigurator/planning/src/blueprint_editor/sled_editor.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_editor/sled_editor.rs]: Thanks for the great comment!

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs (hunk 1 above)
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:892]: Ah, that makes sense. I think it's probably fine to remove the check. We know when we want to expunge something which should trump everything else.

**Comment by andrewjstone** [comment]
Thanks for all the cleanup @jgallagher. Looking good.

**Pull request merged**